Add support for fetch requests and Nextjs static paths and props

ionutpasca opened the pull request.

elizamaria committed:
Merge pull request #742 from teleporthq/chore/update-dangerous-html

(chore): Update dangerous-html package

pascaionut12345 committed:
feat(nextjs static paths and props): nextjs static paths and props

Add the possibility for each page to define resources for static paths and props

pascaionut12345 committed:
feat: fetch support

Add support for fetch requests and nextjs static paths and props

pascaionut12345 committed:
Small change

pascaionut12345 committed:
Merge branch 'development' of github.com:teleporthq/teleport-code-generators into development

pascaionut12345 committed:
Merge branch 'development' into feat/better-fetch

pascaionut12345 committed:
Small fix on template concat

pascaionut12345 committed:
Small change

pascaionut12345 committed:
Get the first item on details page - initial props

pascaionut12345 committed:
Fix tests

ionutpasca commented:
**Make sure to make this one before** -> https://github.com/teleporthq/teleport-code-generators/pull/749

elizamaria commented:
Looks like it's good to go. Should check with @vylly the style/assets support for the nested pages.

ionutpasca committed:
Add support for pagination on static paths/props (#749)

* Add support for pagination on static paths/props

* Small changes in cms project uidl

---------

Co-authored-by: Ionut Pasca <ionut.pasca@teleporthq.io>

pascaionut12345 committed:
Add support for context

pascaionut12345 committed:
Small change

pascaionut12345 committed:
Primitive implementation for context support

pascaionut12345 committed:
Primitive implementation for cms-list and item support

pascaionut12345 committed:
Handle cms-list node individually

pascaionut12345 committed:
Handle cms-list node individually

pascaionut12345 committed:
Remove package.lock

pascaionut12345 committed:
Fix tests

pascaionut12345 committed:
Handle CMS project uidl in standalone test

elizamaria committed:
Merge pull request #760 from teleporthq/context-support

Add CMS list+item nodes support

pascaionut12345 committed:
Add support for CMS list and CMS item inside nextjs generators

vladnicula commented on packages/teleport-project-plugin-contexts/src/next.ts:
```diff
@@ -0,0 +1,51 @@
+export const use${contextName} = () => {
+  const context = useContext(${contextName})
+  if (!context) {
+    throw new Error('use${contextName} must be used within a ${contextName}')
```
```suggestion
    throw new Error('use${contextName} must be used within a ${contextName}Provider')
```

vladnicula commented on packages/teleport-plugin-next-context/src/index.ts:
```diff
@@ -0,0 +1,95 @@
+import {
```
this file is for context providers, not context declaration or read. I was expecting a context read/access not provision in this file. Maybe naming could be better.

vladgrecu committed:
Merge pull request #763 from teleporthq/feat/add-cms-nodes-support

Add support for CMS list and CMS item inside nextjs generators

ionutpasca committed:
Changes required for contentful support (#764)

* Changes required for contentful support

* Small changes

* Add possibility to use mapper functions for cms list/items & initialProps (#767)

Co-authored-by: Ionut Pasca <ionut.pasca@teleporthq.io>

---------

Co-authored-by: Ionut Pasca <ionut.pasca@teleporthq.io>

vladnicula committed:
Feat/dynamic links support for cms details pages (#765)

* Progress so far

* Progress so far

* aRemove type cast that was unnecesarry

* A better example

* Feat/contentful richtext (#768)

* Support for dynamic attributes from ctx

* Test and fix for contentful richtext node

* Remove console.log

* We dont use package.lock

* Bring back standalone

---------

Co-authored-by: Eliza Nitoi <eliza.nitoi@gmail.com>
Co-authored-by: Ionut Pasca <ionut.pasca@teleporthq.io>

ionutpasca committed:
Allow static values on component seo and meta tags (#772)

* Allow static values on component seo and meta tags

* Small change

---------

Co-authored-by: Ionut Pasca <ionut.pasca@teleporthq.io>

TudorCe committed:
Subselector support for classes (#770)

* Subselector support for classes

* Add explanation for &

* Support states and MQs for subselector classes as well

elizamaria committed:
Fix run standalone for next project with cms uidl

vladnicula committed:
Better debug for problematic context providers

vladgrecu committed:
fix/THQ-2027 read-nested-data (#773)

pascaionut12345 committed:
Check if node attributes have contexts

ionutpasca committed:
Check if node attributes have contexts (#774)

Co-authored-by: Ionut Pasca <ionut.pasca@teleporthq.io>

pascaionut12345 committed:
Some changes to fix the contentful normalisation

pascaionut12345 committed:
Merge with better fetch

pascaionut12345 committed:
Revert "Merge with better fetch"

This reverts commit 618b405fd62fa1357966fde046904cbd55df2ad2.

elizamaria committed:
Merge pull request #775 from teleporthq/feat/normalize-cms-content

Feat/normalize cms content

pascaionut12345 committed:
Use id as fallback for context if

TudorCe committed:
Merge pull request #776 from teleporthq/feat/id-or-ctxid

Use id as fallback for context if

JayaKrishnaNamburu commented on examples/uidl-samples/cms-project.json:
```diff
@@ -0,0 +1,3684 @@
+                "resourceMappers": [
+                  {
+                    "name": "resolveContentfulResponse",
+                    "resource": {
```
Can this be changed to `dependency` to keep it consistent with UIDLElemetNode. And second if we can use `importDefinitions` and just make it a reference here. Would also be a option
https://github.com/teleporthq/teleport-code-generators/blob/53c7b8a9ffbe8e18341da464580e08f5292c3f00/examples/uidl-samples/react-project.json#LL181C1-L181C1

ionutpasca replied:
Yes, we can do that

JayaKrishnaNamburu commented on examples/uidl-samples/cms-project.json:
```diff
@@ -0,0 +1,3684 @@
+                    "name": "normalizeContent",
+                    "resource": {
+                      "type": "package",
+                      "version": "git+https://github.com/ionutpasca/contentful-normaliser.git",
```
This can be moved into `teleporthq` organisation. And we can release it. And make the version `latest`. So, the deployments always get the latest normaliser

ionutpasca replied:
Yes...this was only to ship this functionality faster. We can use any package from anywhere :D

JayaKrishnaNamburu commented on examples/uidl-samples/cms-project.json:
```diff
@@ -0,0 +1,3684 @@
+    }
+  },
+  "resources": {
+    "rootFolder": "pages/api",
```
We can make this framework's strategy dependent instead of `pages/api`.

ionutpasca replied:
Yes...this is a good idea

JayaKrishnaNamburu commented on examples/uidl-samples/cms-project.json:
```diff
@@ -0,0 +1,3684 @@
+          "type": "env",
+          "value": "CMS_URL"
+        },
+        "authToken": {
```
We can split this into `headers` and `params`. And each param can have a param type. Like `url_param` and `obj` for post data. Something like that. We can switch the `{ method: 'POST' }` if the params preset as `obj` something like that.

ionutpasca replied:
Yes... we had a discussion about this. We could have a `headers` property inside this object and `authToken` would be inside of it. It remained like this for a faster shipment :)

JayaKrishnaNamburu commented on examples/uidl-samples/cms-project.json:
```diff
@@ -0,0 +1,3684 @@
+        "fileName": "BlogPost"
+      }
+    }
+  }
```
Should we abstract out the whole `params` and `url_params`. We can write a `lib` that takes, `path`, `params` andmake the fetch calls. Instead of generating every time. Something like
``
import fetcher from '@teleporthq/fetcher'

const res = fetcher(path, params)
``

So, just params ad have param type, `token`, `post`, `url_param` something like that

JayaKrishnaNamburu replied:
Just like `axios` but with opinionated for our requirements ?

ionutpasca replied:
This is an option as well, we only had an internal discussion about it. We could do it, but it would mean that the implementation is hidden from the code. We need to decide if this is what we want to do or not. As a developer, I would like to see what and how the fetch request is done, not to be hidden from me; this is a subjective opinion

JayaKrishnaNamburu commented on packages/teleport-plugin-common/src/builders/style-builders.ts:
```diff
@@ -146,11 +157,19 @@ export const generateStylesFromStyleSetDefinitions = (
-            createCSSClassWithSelector(name, `&:${styleRef.meta.state}`, collecedMediaStyles)
+            createCSSClassWithSelector(
+              name,
+              `&${subselectors || ''}:${styleRef.meta.state}`,
```
Can we try using this approach for selectors in react-jss
https://cssinjs.org/jss-plugin-nested?v=v10.6.0

JayaKrishnaNamburu reviewed (commented):
Some improvements we can do, we can extract out something like `fetcher` and some framework specific things. We can extend the project strategy with `contexts` and `resources`. So, they will be handled right with the project generators. Without adding any plugins. Or maybe add those plugins by default to the project-generators. So, by default they will start working.

JayaKrishnaNamburu commented:
Something like this

```js
const generator = createProjectGenerator({
    id: 'teleport-project-next',
    style: ReactStyleVariation.StyledJSX,
    components: {
      ....
    },
    pages: {
      .....
    },
    projectStyleSheet: {
      .....
    },
    entry: {
     .....
    },
    framework: {
      config: {
        ....
      },
      externalStyles: {
        ....
      },
    },
    static: {
     ...
    },
   cms: {
    ....
   }
  })
```

Something like this, so each framework can use it's own plugins to read the paths and things like that.

ionutpasca commented:
>

ionutpasca commented:
Ideally, we should not have any CMS related stuff inside the UIDL. The fact that we added a CMSListItem and a CMSItem node inside the the UIDL was only to speed up the development. We added the `resource` entity to be agnostic from any external dependency that one might use inside the project. we basically needed a `fetch` integration, that is currently used with the CMS, but it could be used in the same time with any external API.

I'm not really sure what you would add inside the `cmd` property that you mentioned above:)

JayaKrishnaNamburu commented:
Hey, i got it. Just bouncing some ideas, if we want the CMS too be inside the UIDL. But, `cms` inside strategy will be allow the pre-defined setups that are needed. Like deps, fetchers, api locations etc. Anything that the framework needs to fetch the data.
